Adding type hints to deep_lift.py

kolvia opened the pull request:
Adding type hints to:
```
captum/attr/_core/deep_lift.py
tests/attr/test_deeplift_basic.py
tests/attr/test_deeplift_classification.py
```

I've added a number of type hints so far but I'm running into issues with what's left. The biggest is that whenever I try to add type hints to `DeepLiftShap.attribute`, I get: `error: Signature of "attribute" incompatible with supertype "DeepLift"`. So, just wanted to get some feedback on what I have so far and see if there are any suggestions going forward.

vivekmig commented on captum/attr/_core/deep_lift.py:
```diff
@@ -55,16 +60,57 @@ def is_output_cloned(output_fn, input_grad_fn):
         self.model = model
-        self.forward_handles = []
-        self.backward_handles = []
+        self.forward_handles: List[Any] = []
```
This is more precisely a List of RemovableHandle (torch.utils.hooks).

vivekmig commented on captum/attr/_core/deep_lift.py:
```diff
@@ -310,10 +356,10 @@ def forward_fn():
         return type(module) in SUPPORTED_NON_LINEAR.keys()
 
-    def _forward_pre_hook_ref(self, module, inputs):
+    def _forward_pre_hook_ref(self, module: Any, inputs: TensorOrTupleOfTensors) -> None:
```
This module should also be Module, note that this may require another "# type: ignore", since we add an attribute to the Module, which is not typed in the Module.

vivekmig commented on tests/attr/test_deeplift_basic.py:
```diff
@@ -263,8 +275,13 @@ def test_lin_maxpool_lin_classification(self):
+        model: Module,
+        attr_method: Union[DeepLift, DeepLiftShap],
+        inputs: Tuple[Tensor, ...],
+        baselines,  # ?
```
You can leave out type hinting this argument, it's a little messy since it's used by both DeepLift and DeepLiftShap, but they don't have consistent baseline arguments.

vivekmig commented:
Also, it looks like there's some lint issues, so once you finish your changes, just run black and flake8 to address linter issues. Make sure these are upgraded to the latest versions, since black may make extraneous changes to other files otherwise.

kolvia commented:
@vivekmig Thanks for the help! 

I've added some more type hinting (and fixed the linter issues) but I'm running into more errors:

`captum/attr/_core/deep_lift.py:499: error: Argument 1 to "register_backward_hook" of "Module" has incompatible type "Callable[[Module[Any], TensorOrTupleOfTensors, TensorOrTupleOfTensors, float], Any]"; expected "Callable[[Module[Any], Union[Tuple[Tensor, ...], Tensor], Union[Tuple[Tensor, ...], Tensor]], Optional[Tensor]]"
captum/attr/_core/deep_lift.py:841: error: Incompatible types in assignment (expression has type "Tuple[Any, ...]", variable has type "Tuple[int, int]")
captum/attr/_core/deep_lift.py:842: error: No overload variant of "mean" matches argument types "Tensor", "int", "bool"
captum/attr/_core/deep_lift.py:842: note: Possible overload variants:
captum/attr/_core/deep_lift.py:842: note:     def mean(self: Tensor, *, dtype: Optional[dtype] = ..., out: Optional[Tensor] = ...) -> Tensor
captum/attr/_core/deep_lift.py:842: note:     def mean(self: Tensor, dim: Union[int, Union[Size, List[int], Tuple[int, ...]]], keepdim: bool = ..., *, dtype: Optional[dtype] = ..., out: Optional[Tensor] = ...) -> Tensor
captum/attr/_core/deep_lift.py:842: note:     def mean(self: Tensor, dim: List[Optional[str]], keepdim: bool = ..., *, dtype: Optional[dtype] = ..., out: Optional[Tensor] = ...) -> Tensor
captum/attr/_core/deep_lift.py:992: error: Argument 1 to "list" has incompatible type "Union[Size, Tensor, Module[Any]]"; expected "Iterable[Any]"
captum/attr/_core/deep_lift.py:1010: error: Argument 1 to "list" has incompatible type "Union[Size, Tensor, Module[Any]]"; expected "Iterable[Any]"
captum/attr/_core/neuron/neuron_deep_lift.py:186: error: Incompatible return value type (got "Union[Tensor, Tuple[Tensor, Tensor]]", expected "Tensor")
captum/attr/_core/neuron/neuron_deep_lift.py:186: error: Incompatible return value type (got "Union[Tuple[Tensor, ...], Tuple[Tuple[Tensor, ...], Tensor]]", expected "Tuple[Tensor, ...]")
captum/attr/_core/neuron/neuron_deep_lift.py:360: error: Argument 2 to "attribute" of "DeepLiftShap" has incompatible type "Union[float, Tensor, Tuple[Union[Tensor, float], ...], Callable[..., Union[Tensor, Tuple[Tensor, ...]]], None]"; expected "Union[Tensor, Callable[..., Tensor]]"
captum/attr/_core/neuron/neuron_deep_lift.py:360: error: Argument 2 to "attribute" of "DeepLiftShap" has incompatible type "Union[float, Tensor, Tuple[Union[Tensor, float], ...], Callable[..., Union[Tensor, Tuple[Tensor, ...]]], None]"; expected "Union[Tuple[Tensor, ...], Callable[..., Tuple[Tensor, ...]]]"
Found 9 errors in 2 files (checked 44 source files)`

I'm kind of stuck on how to fix these. Additionally, there's still some functions I haven't finished adding types to (`_check_valid_module` and the maxpool functions) where I'm not really sure what types they take.

vivekmig commented on captum/attr/_core/deep_lift.py:
```diff
@@ -792,7 +950,14 @@ def maxpool3d(module, inputs, outputs, grad_input, grad_output, eps=1e-10):
+    grad_input,
+    grad_output,
+    eps: float = 1e-10,
 ):
```
This method has two instances of `list(module.input.shape)`, you can replace these with `list(cast(torch.Size, module.input.shape))`

vivekmig commented on captum/attr/_core/deep_lift.py:
```diff
@@ -55,16 +62,57 @@ def is_output_cloned(output_fn, input_grad_fn):
+        self,
+        inputs: TensorOrTupleOfTensors,
+        baselines: Optional[
+            Union[int, Tuple[int, ...], Tensor, List[Tuple[int, ...]]]
```
This baselines hint should match the ones below, `Union[Tensor, int, float, Tuple[Union[Tensor, int, float], ...]]`

kolvia commented:
@vivekmig I think it should be good to go, now. Thanks so much for all your help!

vivekmig commented on captum/attr/_core/deep_lift.py:
```diff
@@ -427,14 +488,16 @@ def _register_hooks(self, module):
-        def pre_hook(module, baseline_inputs_add_args):
+    def _pre_hook_main_model(self) -> RemovableHandle:
+        def pre_hook(
+            module: Module, baseline_inputs_add_args: Tuple[Tensor, ...]
```
Since the tuple includes additional arguments which could be anything, the type of both baseline_inputs_add_args and the return type should be just Tuple.

vivekmig commented on tests/attr/test_deeplift_basic.py:
```diff
@@ -209,8 +213,12 @@ def gen_baselines_returns_array():
+        def custom_attr_func(
+            multipliers: Tuple[Tensor, ...],
+            inputs: Tuple[Tensor, ...],
+            baselines: Tuple[Union[Tensor, int, float], ...],
```
This type hint for baselines should actually be Tuple[Tensor, ...], since we internally convert to a Tensor before this method is called.
